File: 981.cpp
#include <algorithm>
#include <deque>
#include <iostream>
#include <map>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
class TimeMap {
  struct Val {
    string value;
    int timestamp;
    Val(string value_in, int timestamp_in)
        : value(value_in), timestamp(timestamp_in) {}
  };

public:
  TimeMap() {}

  unordered_map<string, vector<Val>> m;
  void set(string key, string value, int timestamp) {
    m[key].emplace_back(value, timestamp);
  }

  string get(string key, int timestamp) {
    auto it = m.find(key);
    if (it == m.end()) {
      return "";
    }
    auto nums = it->second;
    int l = 0, r = nums.size() - 1;
    while (l <= r) {
      int m = (r - l) / 2 + l;
      if (nums[m].timestamp < timestamp) {
        l = m + 1;
      } else if (nums[m].timestamp > timestamp) {
        r = m - 1;
      } else {
        return nums[m].value;
      }
    }
    if (r >= 0) {
      return nums[r].value;
    }
    return "";
  }
};

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap* obj = new TimeMap();
 * obj->set(key,value,timestamp);
 * string param_2 = obj->get(key,timestamp);
 */
void test1() {
  TimeMap t;
  t.set("love", "high", 10);
  t.set("love", "low", 20);
  t.get("love", 5);
  t.get("love", 10);
  t.get("love", 15);
  t.get("love", 20);
  t.get("love", 25);
}
void test2() {
  TimeMap t;
  t.set("love", "high", 10);
  t.set("love", "low", 20);
  cout << t.get("love", 5) << endl;
  cout << t.get("love", 10) << endl;
  cout << t.get("love", 15) << endl;
  cout << t.get("love", 20) << endl;
  cout << t.get("love", 25) << endl;
}
int main() {
  test1();
  test2();
}

File: 981.py
class TimeMap:
    def __init__(self):
        self.m = {}

    def set(self, key: str, value: str, timestamp: int) -> None:
        if key in self.m:
            self.m[key].append((timestamp, value))
        else:
            self.m[key] = [(timestamp, value)]
        return None

    def get(self, key: str, timestamp: int) -> str:
        if key not in self.m:
            return ""
        nums = self.m[key]
        l, r = 0, len(nums) - 1
        while l < r:
            m = (r - l) // 2 + l
            if nums[m][0] > timestamp:
                r = m
            elif nums[m][0] < timestamp:
                l = m + 1
            else:
                return nums[m][1]
        while l > 0 and nums[l][0] > timestamp:
            l -= 1
        if nums[l][0] > timestamp:
            return ""
        return nums[l][1]


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)

t = TimeMap()
# print(t.set("foo", "bar", 1))
# print(t.get("foo", 1))
# print(t.get("foo", 3))
# print(t.set("foo", "bar2", 4))
# print(t.get("foo", 4))
# print(t.get("foo", 5))

print(t.set("love", "high", 10))
print(t.set("love", "low", 20))
print(t.get("love", 5))
print(t.get("love", 10))
print(t.get("love", 15))
print(t.get("love", 20))
print(t.get("love", 25))
